containers/unit_tests/CMakeLists.txt

```diff
@@ -3,108 +3,32 @@ KOKKOS_INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR})
-    serial/TestSerial_Vector.cpp
-    serial/TestSerial_ViewCtorPropEmbeddedDim.cpp
-  )
-ENDIF()
-
-IF(Kokkos_ENABLE_OPENMP)
-KOKKOS_ADD_EXECUTABLE_AND_TEST(
-  UnitTest_OpenMP
-  SOURCES
-    UnitTestMain.cpp
-    openmp/TestOpenMP_BitSet.cpp
-    openmp/TestOpenMP_DualView.cpp
-    openmp/TestOpenMP_DynamicView.cpp
-    openmp/TestOpenMP_DynRankViewAPI_generic.cpp
-    openmp/TestOpenMP_DynRankViewAPI_rank12345.cpp
-    openmp/TestOpenMP_DynRankViewAPI_rank67.cpp
-    openmp/TestOpenMP_ErrorReporter.cpp
-    openmp/TestOpenMP_OffsetView.cpp
-    openmp/TestOpenMP_ScatterView.cpp
-    openmp/TestOpenMP_StaticCrsGraph.cpp
-    openmp/TestOpenMP_UnorderedMap.cpp
-    openmp/TestOpenMP_Vector.cpp
-    openmp/TestOpenMP_ViewCtorPropEmbeddedDim.cpp
-  )
-ENDIF()
-
-IF(Kokkos_ENABLE_HPX)
-KOKKOS_ADD_EXECUTABLE_AND_TEST(
-  UnitTest_HPX
-  SOURCES
-    UnitTestMain.cpp
-    hpx/TestHPX_BitSet.cpp
-    hpx/TestHPX_DualView.cpp
-    hpx/TestHPX_DynamicView.cpp
-    hpx/TestHPX_DynRankViewAPI_generic.cpp
-    hpx/TestHPX_DynRankViewAPI_rank12345.cpp
-    hpx/TestHPX_DynRankViewAPI_rank67.cpp
-    hpx/TestHPX_ErrorReporter.cpp
-    hpx/TestHPX_OffsetView.cpp
-    hpx/TestHPX_ScatterView.cpp
-    hpx/TestHPX_StaticCrsGraph.cpp
-    hpx/TestHPX_UnorderedMap.cpp
-    hpx/TestHPX_Vector.cpp
-    hpx/TestHPX_ViewCtorPropEmbeddedDim.cpp
-  )
-ENDIF()
-
-IF(Kokkos_ENABLE_CUDA)
-KOKKOS_ADD_EXECUTABLE_AND_TEST(
-  UnitTest_Cuda
-  SOURCES
-    UnitTestMain.cpp
-    cuda/TestCuda_BitSet.cpp
-    cuda/TestCuda_DualView.cpp
-    cuda/TestCuda_DynamicView.cpp
-    cuda/TestCuda_DynRankViewAPI_generic.cpp
-    cuda/TestCuda_DynRankViewAPI_rank12345.cpp
-    cuda/TestCuda_DynRankViewAPI_rank67.cpp
-    cuda/TestCuda_ErrorReporter.cpp
-    cuda/TestCuda_OffsetView.cpp
-    cuda/TestCuda_ScatterView.cpp
-    cuda/TestCuda_StaticCrsGraph.cpp
-    cuda/TestCuda_UnorderedMap.cpp
-    cuda/TestCuda_Vector.cpp
-    cuda/TestCuda_ViewCtorPropEmbeddedDim.cpp
-  )
-ENDIF()
-
+foreach(Tag Threads;Serial;OpenMP;HPX;Cuda)
+  # Because there is always an exception to the rule
+  if(Tag STREQUAL "Threads")
+    set(ARCH "PTHREAD")
+  else()
+    string(TOUPPER ${Tag} ARCH)
+  endif()
+  # Add test for that backend if it is enabled
+  if(Kokkos_ENABLE_${ARCH})
+    KOKKOS_ADD_EXECUTABLE_AND_TEST(
+      UnitTest_${Tag}
+      SOURCES
+        UnitTestMain.cpp
+        ${Tag}/Test${Tag}_BitSet.cpp
+        ${Tag}/Test${Tag}_DualView.cpp
+        ${Tag}/Test${Tag}_DynamicView.cpp
+        ${Tag}/Test${Tag}_DynRankViewAPI_generic.cpp
+        ${Tag}/Test${Tag}_DynRankViewAPI_rank12345.cpp
+        ${Tag}/Test${Tag}_DynRankViewAPI_rank67.cpp
+        ${Tag}/Test${Tag}_ErrorReporter.cpp
+        ${Tag}/Test${Tag}_OffsetView.cpp
+        ${Tag}/Test${Tag}_ScatterView.cpp
+        ${Tag}/Test${Tag}_StaticCrsGraph.cpp
+        ${Tag}/Test${Tag}_UnorderedMap.cpp
+        ${Tag}/Test${Tag}_Vector.cpp
+        ${Tag}/Test${Tag}_ViewCtorPropEmbeddedDim.cpp
+      )
+  endif()
+endforeach()
```

Comment:
@jjwilke This looks good to you?

Reply:
This looks fine to me. I'm a bit curious why we need separate source files for the tests? Is this not something where we could `-D${default_backend}` and run the same source code for each?

Reply:
Yes that's how I would do it but I don't want to deal with the generated makefiles and I suppose we are not dropping the support yet.

Reply:
Uhm: isn't TEST_CATEGORY and TEST_EXECSPACE different in the different cpp files, since they include different header files for the TestCategory? I mean I suppose we could add the defines to the compiler line, but generally I don't like that approach ...

Reply:
Oh I see @jjwilke, you didn't mean a cmake option -D ... but a define on the compile line? We can't just change the default backend, because the tests need to explicitly test whether stuff works with say OpenMP when OpenMP is not the default backend (because you had CUDA enabled).